modules/layers/src/path-layer/tesselator.js

```diff
@@ -0,0 +1,159 @@
+    if (newSize <= typedArray.length) {
+      return typedArray;
+    }
+
+    // Allocate at least one element to ensure a valid buffer
+    const allocSize = Math.max(Math.ceil(newSize * this.overAlloc), 1);
+    const newArray = this.allocate(typedArray.constructor, allocSize);
+
+    if (copy) {
+      newArray.set(typedArray);
+    }
+
+    this.release(typedArray);
+    return newArray;
+  }
+
+  release(typedArray) {
+    // TODO - add to pool
+    // logFunctions.onUpdate({
+    //   level: LOG_DETAIL_PRIORITY,
+    //   message: `${attributeName} allocated ${allocCount}`,
+    //   id: this.id
+    // });
+  }
+}
+
+export default class Tesselator {
+  constructor(opts = {}) {
+    const {attributes = {}} = opts;
+
+    this.typedArrayManager = new TypedArrayManager();
+    this.bufferLayout = null;
+    this.instanceCount = 0;
+    this.attributes = {};
+    this._attributeDefs = attributes;
+
+    this.updateGeometry(opts);
+
+    Object.seal(this);
+  }
+
+  /* Public methods */
+  updateGeometry({data, getGeometry, fp64}) {
+    this.data = data;
+    this.getGeometry = getGeometry;
+    this.fp64 = fp64;
+    this._rebuildGeometry();
+  }
+
+  /* Subclass interface */
+
+  // Fill the typed arrays for geometry attributes
+  updateGeometryAttributes() {
+    throw new Error('Not implemented');
+  }
+
+  // Returns the number of vertices in a geometry
+  countInstancesInGeometry(geometry) {
+    throw new Error('Not implemented');
+  }
+
+  /* Private utility methods */
+
+  /**
+   * Visit all objects
+   * `data` is expected to be an iterable consistent with the base Layer expectation
+   */
+  _forEachGeometry(visitor) {
+    const {data, getGeometry} = this;
+    let dataIndex = 0;
+    for (const object of data) {
+      const geometry = getGeometry(object);
+      visitor(geometry, dataIndex++);
+    }
+  }
+
+  _updateAttribute({target, size, getValue}) {
+    const {data, bufferLayout} = this;
+
+    let i = 0;
+    let dataIndex = 0;
+    for (const object of data) {
+      const value = getValue(object, dataIndex);
+      const numVertices = bufferLayout[dataIndex++];
+
+      fillArray({target, source: value, start: i, count: numVertices});
+      i += numVertices * size;
+    }
+    return target;
+  }
+
+  _rebuildGeometry() {
+    if (!this.data || !this.getGeometry) {
+      return;
+    }
+
+    // count instances
+    const bufferLayout = [];
+    let instanceCount = 0;
+    this._forEachGeometry((geometry, dataIndex) => {
+      const count = this.countInstancesInGeometry(geometry);
+      instanceCount += count;
+      bufferLayout[dataIndex] = count;
+    });
+
+    // allocate attributes
+    const {attributes, _attributeDefs, typedArrayManager, fp64} = this;
+    for (const name in _attributeDefs) {
+      const def = _attributeDefs[name];
+
+      if (!def.fp64 || fp64) {
+        attributes[name] = attributes[name]
+          ? typedArrayManager.reallocate(attributes[name], def.size * instanceCount)
```

Comment:
Maybe just make `reallocate` fall back to `allocate` if supplied vector is `null`?